control-plane/api-gateway/binding/cleanup.go

```diff
@@ -0,0 +1,198 @@
+package binding
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	mapset "github.com/deckarep/golang-set/v2"
+	"github.com/go-logr/logr"
+	"github.com/hashicorp/consul-k8s/control-plane/consul"
+	"github.com/hashicorp/consul/api"
+)
+
+const (
+	oldACLRoleName   = "managed-gateway-acl-role"
+	oldACLPolicyName = "api-gateway-token-policy"
+)
+
+var sleepTime = 60 * time.Second
+
+type Cleaner struct {
+	Logger       logr.Logger
+	ConsulConfig *consul.Config
+	ServerMgr    consul.ServerConnectionManager
+	AuthMethod   string
+}
+
+// Run periodically cleans up old ACL roles and policies as well as orphaned inline certs
+// when it detects that there are no more inline certs and the old ACL role and policy are not in use then it exits.
+func (c Cleaner) Run(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(sleepTime):
+		}
+
+		client, err := consul.NewClientFromConnMgr(c.ConsulConfig, c.ServerMgr)
+		if err != nil {
+			c.Logger.Error(err, "failed to create Consul client")
+			continue
+		}
+
+		aclsCleanedUp, err := c.cleanupACLRoleAndPolicy(client)
+		if err != nil {
+			c.Logger.Error(err, "failed to cleanup old ACL role and policy")
+		}
+
+		inlineCertsAllCleanedUp, err := cleanupInlineCerts(client)
+		if err != nil {
+			c.Logger.Error(err, "failed to cleanup inlineCerts")
+		}
+
+		if aclsCleanedUp && inlineCertsAllCleanedUp {
+			return
+		}
+	}
+}
+
+// cleanupACLRoleAndPolicy deletes the old shared gateway ACL role and policy if they exist.
+func (c Cleaner) cleanupACLRoleAndPolicy(client *api.Client) (bool, error) {
+	existingRules, _, err := client.ACL().BindingRuleList(c.AuthMethod, &api.QueryOptions{})
+	if err != nil {
+		if err.Error() == "Unexpected response code: 401 (ACL support disabled)" {
+			return true, nil
+		}
+		return false, fmt.Errorf("failed to list binding rules: %w", err)
+	}
+
+	oldBindingRules := make(map[string]*api.ACLBindingRule)
+
+	// here we need to find binding rules with the old name that have a matching selector to the new gateway specific binding rule
```

Comment:
so another direction we could've gone down here is to fetch all the gateways and then for each gateway match it to a binding rule and if we match each gateway to a new binding rule then we can safely delete all the binding rules, definitely open to push back to change this to utilize that heuristic for deleting binding rules rather than what we have here (and we already fetch all the gateways for the inline cert cleanup so it's not like we don't need that data anyways)